### ISSUE TITLE
DECORATE -> ZScript migration

### ISSUE THREAD

**Issue by Username-N00b-is-not-available** [Realm667/WolfenDoom#470]
I can't think of a worse time moment than now to do such large perturbations in the game, but I think this is quite a serious problem, and I doubt we would do anything with it after the release.
My suggestion is to move **all** the DECORATE code we use in BoA to ZScript. I know it is more than a megabyte, and nobody would help me in this, so **I am willing to do it myself**. I am not going to refactor or rewrite anything, but rather move all files almost verbatim, only doing the necessary syntax changing.
The main reason for this change is that you cannot inherit ZScript actors from DECORATE ones, and this limits the modding capabilities severely (which I've learnt for myself when making mutators #459). To add a new function to actor (make an MP40Guard throw grenades, say) you need to copy its whole definition (including 2 to 4 parent classes exposed to DECORATE), this leads to awful amounts of copy-pasting. Every simple addon which changes a few DECORATE actors becomes a difficult to manage mess.
***
I would just like to ask if my changes could be tolerated, and possibly also ask for comments on how to do this better. I am thinking of doing this in a single huge commit. @AFADoomer @Ozymandias81 @Talon1024

**Comment by Tormentor667** [comment]
Even though you didn’t ask me in particular, your thoughts make sense and I agree (in terms of future mods) that it makes sense to consider such a change.

I am just not sure how dangerous it is to mess with that now.

**Comment by Ozymandias81** [comment]
I also had this in mind during the development, even AFADoomer and Talon1024 (more from themselves, since I am not good with ZScript), but we had some issues to take in consideration:

1 - Pick a stable build of GZDoom and consider that in the nearly future NashShadows will be merged into GZDoom code itself
2 - The blood system works in a peculiar way that indeed might be improved via zscript (droplets and old nashgore code), but I did some workarounds to have translated gibs/effects work properly over some "special" solutions which works fine via decorate but that might need a total rework via zscript (and we must consider that actual mix has some performance impact due to some values that needs to be set as absolute imho, too many options for players "blood" wise doesn't make sense)
3 - Most of my work will become totally unrecognizable and this will also mean that my 6 years of work will disappear in the air like a Thanos snap
4 - We must do something to improve the cull system and actors in general since these terrible spikes of slodowns I get might not come from gzdoom devbuilds, but the project itself. This is way more important than anything else.

Doing that right now before release would be painful and a total no from me. Regarding the future well, not sure if I'll continue modding, too many things have changed on BoA from the original idea and some got worse somehow, too many features all together and too many stuff on screen ongoing, the main thing behind BoA was "don't get overboard", that's how it used to be and personally I felt very overburnt the last year, but hey who am I to say "no" if the project (alongside with the engine itself) will improve in the future?

**Comment by Talon1024** [comment]
I do not think it is a good idea to make such a huge change when we are right on the proverbial doorstep of the final release. It could cause a huge amount of bugs, regressions, and unexpected fundamental changes (like the damage done by the tank cannon projectiles, since they use the Damage property, and ZScript introduced the DamageFunction property).

If you wish to convert all of the DECORATE code to ZScript, please do so in a separate branch, and don't merge it into master until we release the final version.

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 @Ozymandias81 @Talon1024 Thank you for attention to this issue! I would like to post just one message with @mentions to clarify the OP.
As I said previously, this suggestion is not about 'cleaning up', or 'refactoring', or modifying the code in any substantial way; it is only about **conversion** (so points 1, 2, 4 of @Ozymandias' post are only tangentially related). Most of the process is going to be purely mechanical semicolon addition :-D, and I promise I won't use any automated tools to do it. I will check every file, line by line, myself (if I manage to do this all in time, of course; if not, then I simply won't commit it).

@Talon1024, as for Damage/DamageFunction-type inconsistencies, I know about them, and there is even a guide on the wiki https://zdoom.org/wiki/Converting_DECORATE_code_to_ZScript#Required_changes. Damage won't even compile for non-constant expressions anymore. There are very few differences between DECORATE and ZScript, but if I stumble upon something that looks suspicious, I will research into it myself, or at least ask you here.

@Ozymandias81, I don't think adding a few semicolons after the codepointers could make your efforts unrecognizable, but that is an important topic. It would be good to add license headers to DECORATE files (and not only them, but I shall stay on topic), similarly to how @AFADoomer and @Talon1024 do in ZScript ones. Just choose (or write) your license :-) I'll be able to add it then.

***

Lastly, I'm concerned about this so much now, because
1. I just know we aren't ever going to fix this after the release. @AFADoomer and @Talon1024 have little time. @Tormentor667 had already announced his departure. Now @Ozymandias81 writes that he is unsure whether to continue modding or not. Actually, I'm not really sure for myself, too, and I want to give our team and the Doom community the best I could do before retiring myself.
2. (response to @Talon1024's last phrase) Blade of Agony has been in development for around six years. The community has been waiting for almost four years for the final release. The version we are releasing on 30th will be 'The Blade of Agony', the one which will be remembered in the community for years. I doubt too many people would bother redownloading the game after a few months if something so 'miniscule' (from the gameplay perspective) as this is changed, so it would mean the presence of different 'official' versions, hence ugly, overcomplicated and incompatible mods.

Of course, the best way not to introduce new bugs would be to do nothing and stare at the clock and calendar until the release time comes.

**Comment by Tormentor667** [comment]
First of all, everything @Username-N00b-is-not-available says is valid, and I totally support it in general.

The problem for me is, that I can't decide something like this as I am not the one who is in charge of the code or a good coder anyway. I have no idea how dangerous it is to convert the DECORATE to ZScript - and how much benefit we have from that. 

I simply need to ask myself, what's more worth: Cleaning up the code and making work easier for modders or keeping what we have and be sure that we do not introduce new bugs.

_Regarding the future: I will still be open to improve BoA - so adding more functionality and fixing bugs after the release on April 30th is possible if we all want to continue on this. I have no problem with releasing a v3.1. or a v3.2 with such improvements that you can currently fork if you want to, but that's just a side note._

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667

> Cleaning up the code and making work easier for modders

This issue is not about cleaning up the code, well, not in the exact sense. The BoA code will stay as messy as it is now, but just in ZScript :-) It will help modders because **they** will be able to write a lot less messy code.

It would have been very difficult for me to introduce bugs with this procedure :-) I simply need to traverse all the directories and change every file syntax to ZScript one (not counting the subtleties @Talon1024 mentioned here, but they are quite rare, and all possible cases are known and documented in the wiki).
Some weeks ago I skimmed all ZScript files for interesting classes to add to the editing docs, it took me around an hour or two, honestly I don't remember. We have around the same amount of code in ZScript as in DECORATE, so this conversion will take a bit longer, but I regard it as manageable.

**Comment by Tormentor667** [comment]
Isn't there a DECORATE to ZSCRIPT converter somewhere in the ZDoom forums anyway? I am pretty sure that Graf Zahl has written something like this. Automating this with a proper tool would be a less error-prone variant.

**Comment by AFADoomer** [comment]
Converting the DECORATE code to ZScript _properly_ is not the same as just adding semi-colons and a Default {} header.  

The automated tools will do basic conversion that are OK, but still have all of the DECORATE hacks in place.  It'll still work, but if you're going to convert everything, it should be done right. 

That said, a crappy conversion would at least have everything in ZScript so that we could plug away at updating it for a 3.1 or 3.2 release

**Comment by Username-N00b-is-not-available** [comment]
> a crappy conversion would at least have everything in ZScript so that we could plug away at updating it for a 3.1 or 3.2 release

Seconding this very much!

> Isn't there a DECORATE to ZSCRIPT converter somewhere in the ZDoom forums anyway?

Google says that [there is one](https://forum.zdoom.org/viewtopic.php?f=44&t=60380), but it was not written by Graf Zahl, and there is a list of bugs in the thread the converter author had not addressed (the last bug posts are from 2019, and the OP edit date is 2018). I don't think we can trust this program at all. (Not sure if the team trusts myself, but I promise I will do my best.)

**Comment by Tormentor667** [comment]
Well, then let's do this manually and prepare it for a 3.1 release - if you want to start working on it already, you could open a fork until then. Glad that we could solve this :)

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 I am very sorry, but I think you have not understood us properly... What I wanted to say (and from what I read @AFADoomer supports me in it) is that it is better to do the 'verbatim' migration right now, but leave the complete overhaul (removing ugly DECORATE hacks from the resulting code, refactoring it properly etc.) to 3.1. My original take on the issue was only related to the first part, and I am not going to commit **any** changes in code logic before the release.
The whole damn thing is that GZDoom only reads DECORATE definitions after all ZScript ones have been processed, so you can base DECORATE actors on ZScript ones, _but not vice versa_. This is a fundamental limitation of the engine, so there is only this radical way of conversion to combat it.
I'll probably start working on this manually in a few hours anyway.

**Comment by Tormentor667** [comment]
Ah, now I got it. Well, as I said - I am fine with it as long as we can guarantee to not introduce new bugs to the final release. We maybe won't have enough time to do a full test of Chapter 1-3 before the release (even though I think I could handle that if you are quick enough with the conversion).

I'd leave the last word to @Talon1024 @AFADoomer @Ozymandias81 though.

**Comment by AFADoomer** [comment]
As long as we can easily identify the converted code that still may need a final review/rewrite, I don't have a problem with someone starting the conversions.  Fully testing before release might be a problem, though.

**Comment by Username-N00b-is-not-available** [comment]
Folders ```critters``` to ```models``` are done. That's only 291KB out of >1MB, but I think these were the worst part because of heaps of small actor definitions. God, do I hate these one-frame decorations now :-)

**Comment by Tormentor667** [comment]
You are fast as hell

**Comment by Username-N00b-is-not-available** [comment]
I just hope tomorrow I won't receive a Darwin award for this speed :-)

**Comment by DoomJedi** [comment]
Impressive :)

Don't forget the mutators though :)

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi This issue was actually inspired by the mutators, I got too lazy copy-pasting whole DECORATE actor definitions and converting them to ZScript to add just a single line of code. So I had to think of a way to make the future modders' (and my own, of course) life easier.

**Comment by AFADoomer** [comment]
Might be good to rename all of the zscript files from *.txt to *.zs while you are doing this also...

**Comment by Username-N00b-is-not-available** [comment]
I seem to have done everything necessary in https://github.com/Realm667/WolfenDoom/pull/478/commits/f5e0e4111610c2e6b121f211f472283a1fc13e61, only testing is left. **All commits present on ```master``` have been incorporated into this branch manually.**
@Tormentor667 @Ozymandias81 @AFADoomer @Talon1024, please do not make **any zscript/decorate code changes** to the repository in the following few hours (I will be testing maps for any broken things).

**Comment by Tormentor667** [comment]
Hopefully there won't be a need to touch DECORATE from my side - I let people work on that who understand the matter :)

**Comment by Username-N00b-is-not-available** [comment]
I would like to thank everyone for participation in this!
@Tormentor667 @Ozymandias81 @AFADoomer @Talon1024

**Comment by Tormentor667** [comment]
Thanks to all of you for doing this - let’s hope we see some mods now :)

**Comment by AFADoomer** [comment]
Here are some follow-up items (displayed won startup when developer CVAR set to 2):
Script warning, "boa.ipk7:scripts/decorate/props/bath.zs" line 152:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/props/egypt.zs" line 312:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/props/egypt.zs" line 312:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/props/lights.zs" line 2018:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/props/lights.zs" line 2018:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/nazis.zs" line 2967:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/nazis.zs" line 2968:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/nazis.zs" line 2969:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/animals.zs" line 397:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/animals.zs" line 508:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/animals.zs" line 592:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 602:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 610:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 618:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 627:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 635:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 643:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 887:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 895:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 903:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 912:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 920:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 928:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/bosses.zs" line 1116:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/occult.zs" line 260:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/occult.zs" line 263:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/stealth.zs" line 51:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/monsters/stealth.zs" line 73:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 677:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 678:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 685:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 687:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 689:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 691:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 693:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 711:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 712:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 719:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 721:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 723:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 725:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 743:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 744:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 751:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 753:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 755:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 792:
Truncation of floating point value
Script warning, "boa.ipk7:scripts/decorate/misc/droplets.zs" line 800:
Truncation of floating point value

All of these probably just require the floating point value to be cast to an int - 'int(<whatever the double/float value is>)'

**Comment by Username-N00b-is-not-available** [comment]
@AFADoomer Thank you, and I would just like to remind you of the [CKSpike problem](https://github.com/Realm667/WolfenDoom/pull/478#issuecomment-819479405), it is quite important.
@Tormentor667 Definitely ;-)
